packages/tools/client-debugger/src/FluidClientDebugger.ts

```diff
@@ -0,0 +1,354 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { TypedEventEmitter } from "@fluidframework/common-utils";
+import {
+    AttachState,
+    IAudience,
+    IContainer,
+    ICriticalContainerError,
+} from "@fluidframework/container-definitions";
+import { ConnectionState } from "@fluidframework/container-loader";
+import { IFluidLoadable } from "@fluidframework/core-interfaces";
+import { IResolvedUrl } from "@fluidframework/driver-definitions";
+import { IClient, ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+import { MemberChangeKind } from "./Audience";
+import { IFluidClientDebugger, IFluidClientDebuggerEvents } from "./IFluidClientDebugger";
+import { AudienceChangeLogEntry, ConnectionStateChangeLogEntry } from "./Logs";
+
+/**
+ * {@link IFluidClientDebugger} implementation.
+ *
+ * @remarks This class is not intended for external use. Only its interface is exported by the library.
+ *
+ * @internal
+ */
+export class FluidClientDebugger
```

Comment:
i question the architecture here. why do we need to wrap everything, can't the higher level constructs just take the real abstractions, rather than wrapping everything. its add a ton of complexity and long term maintenance which isn't obviously necessary

Reply:
It is a path I am considering. The primary reasons I opted for this pattern to start are:
1. Since our primary driver for this library is to power tooling to accompany our 3rd party offering, I didn't want to expose all of our lower-level concepts outright. This was a means to scope accessibility only to data we explicitly want to expose globally.
2. It records additional historical data (audience changes, etc.), and is a convenient place to do so.

The level of boilerplate has exceeded what I had hoped would be needed, so I do think some changes are needed here. A thinner adapter that just offers restricted access to the underlying container + historical logs might be the right compromise there.

Reply:
does the code matter here? will users ever interact with it, is it even exported? i thought this would all be driving gui we ship, so we don't really need to worry about the internal abstractions

Reply:
It's a fair question. For now, the GUI is the only driver of this, but I could see a tool like this powering other experiences as well. So, for now, we probably don't need to expose anything beyond the initialization / cleanup logic, but I expect we'll want to consider scoping at some point down the line.